tests/Integration/Schema/Directives/UpsertDirectiveTest.php

```diff
@@ -198,6 +199,150 @@ interface IUser
         ]);
     }
 
+    public function testDirectUpsertByIdentifyingColumn(): void
+    {
+        $this->schema .= /** @lang GraphQL */ '
+        type User {
+            id: ID!
+            email: String!
+            name: String!
+        }
+
+        type Mutation {
+            upsertUser(name: String!, email: String!): User @upsert(identifyingColumns: ["email"])
+        }
+        ';
+
+        $this->graphQL(/** @lang GraphQL */ '
+        mutation {
+            upsertUser(
+                email: "foo@te.st"
+                name: "bar"
+            ) {
+                name
+                email
+            }
+        }
+        ')->assertJson([
+            'data' => [
+                'upsertUser' => [
+                    'email' => 'foo@te.st',
+                    'name' => 'bar',
+                ],
+            ],
+        ]);
+
+        $user = User::firstOrFail();
+
+        $this->assertSame('bar', $user->name);
+        $this->assertSame('foo@te.st', $user->email);
+
+        $this->graphQL(/** @lang GraphQL */ '
+        mutation {
+            upsertUser(
+                email: "foo@te.st"
+                name: "foo"
+            ) {
+                name
+                email
+            }
+        }
+        ')->assertJson([
+            'data' => [
+                'upsertUser' => [
+                    'email' => 'foo@te.st',
+                    'name' => 'foo',
+                ],
+            ],
+        ]);
+
+        $user->refresh();
+
+        $this->assertSame('foo', $user->name);
+        $this->assertSame('foo@te.st', $user->email);
+    }
+
+    public function testDirectUpsertByIdentifyingColumns(): void
+    {
+        $company = factory(Company::class)->create(['id' => 1]);
+
+        $this->schema
+            /** @lang GraphQL */
+            .= '
+        type User {
+            id: ID!
+            email: String!
+            name: String!
+            company_id: ID!
```

Comment:
No raw foreign keys in GraphQL, not even as a test case. Often, those test cases serve as an example, so I don't want them to implement anti-patterns.